server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -139,7 +139,9 @@ public void startReplication(ActionListener<Void> listener) {
         final StepListener<GetSegmentFilesResponse> getFilesListener = new StepListener<>();
         final StepListener<Void> finalizeListener = new StepListener<>();
 
+        logger.trace("[shardId {}] Replica starting replication [id {}]", shardId().getId(), getId());
         // Get list of files to copy from this checkpoint.
+        state.setStage(SegmentReplicationState.Stage.GET_CHECKPOINT_INFO);
         source.getCheckpointMetadata(getId(), checkpoint, checkpointInfoListener);
 
         checkpointInfoListener.whenComplete(checkpointInfo -> getFiles(checkpointInfo, getFilesListener), listener::onFailure);
```

Comment:
I think there is an issue between stage demarcation and handlng the unhappy path. @kartg please correct me if I am wrong but in case of failure `checkpointInfoListener` (and others) would delegate to `listener::onFailure` without marking the current stage as "finished", correct?

Reply:
Yes, that's a conscious choice. We're treating the `DONE` stage as a successful replication run. I opted to avoid an explicit `FAILED` stage since the [target](https://github.com/opensearch-project/OpenSearch/blob/8236446666fbe557e256bc38f14a3c6dd8ba76c8/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java#L188) and [state](https://github.com/opensearch-project/OpenSearch/blob/15da49fe723426a1aecbddd8d616b526c675f163/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java#L67) objects are local to each replication, and the replication stages are currently only relevant when tracking timing data

Reply:
That's fine (I think), the dangling stage is what concerns me: since we publish timing on success only, we do nothing on failure (which could be even more important than success for troubleshooting). In any case, leaving the decision up to you.